test/extensions/filters/http/grpc_json_transcoder/grpc_json_transcoder_integration_test.cc

```diff
@@ -331,6 +331,112 @@ TEST_P(GrpcJsonTranscoderIntegrationTest, InvalidJson) {
       R"({ "theme"  "Children" })", {}, {}, Status(),
       Http::TestHeaderMapImpl{{":status", "400"}, {"content-type", "text/plain"}},
       "Expected : between key:value pair.\n", false);
+
+  testTranscoding<bookstore::CreateShelfRequest, bookstore::Shelf>(
+      Http::TestHeaderMapImpl{{":method", "POST"}, {":path", "/shelf"}, {":authority", "host"}},
+      R"({ "theme" : "Children" }EXTRA)", {}, {}, Status(),
+      Http::TestHeaderMapImpl{{":status", "400"}, {"content-type", "text/plain"}},
+      "Parsing terminated before end of input.\n", false);
+}
+
+std::string generateDeepJson(int level, bool valid) {
+  std::string begin = R"({"k":)";
+  std::string deep_val = R"("v")";
+  std::string end = R"(})";
+  std::string json;
+
+  for (int i = 0; i < level; ++i) {
+    absl::StrAppend(&json, begin);
+  }
+  if (valid) {
+    absl::StrAppend(&json, deep_val);
+  }
+  for (int i = 0; i < level; ++i) {
+    absl::StrAppend(&json, end);
+  }
+  return json;
+}
+
+std::string generateDeepProto(int level) {
+  std::string proto = R"(fields{key:"k"value{string_value:"v"}})";
+  for (int i = 0; i < level - 1; ++i) {
+    proto = absl::StrCat(R"(fields{key:"k"value{struct_value{)", proto, R"(}}})");
+  }
+  return proto;
+}
+
+TEST_P(GrpcJsonTranscoderIntegrationTest, DeepStruct) {
+  std::string deepJson = generateDeepJson(32, true);
+  std::string deepProto = "content {" + generateDeepProto(32) + "}";
+  // Due to the limit of protobuf util, we can only compare to level 32.
+  testTranscoding<bookstore::EchoStructReqResp, bookstore::EchoStructReqResp>(
+      Http::TestHeaderMapImpl{
+          {":method", "POST"}, {":path", "/echoStruct"}, {":authority", "host"}},
+      deepJson, {deepProto}, {deepProto}, Status(),
+      Http::TestHeaderMapImpl{
+          {":status", "200"}, {"content-type", "application/json"}, {"grpc-status", "0"}},
+      R"({"content":)" + deepJson + R"(})");
+
+  // The valid deep struct is parsed successfully.
+  // Since we didn't set the response, it return 503.
+  testTranscoding<bookstore::EchoStructReqResp, bookstore::EchoStructReqResp>(
+      Http::TestHeaderMapImpl{
+          {":method", "POST"}, {":path", "/echoStruct"}, {":authority", "host"}},
+      generateDeepJson(100, true), {}, {}, Status(),
```

Comment:
why 100,  not 32?

Reply:
maybe we should send response to make sure 100 level can be parsed and converted successfully.

Reply:
https://github.com/cloudendpoints/esp/blob/42ae4da66a694373b77e3a5dab3da9c406cc08c3/src/nginx/t/transcoding_deep_struct.t#L100
Like esp, for level 100, we only ensure that the proxy can finish the parsing but don't verify the test output.
When the level goes up more than 32, it seems there are some issues between json -> proto or proto -> json.